docs/latest/design/ratelimit.md

```diff
@@ -0,0 +1,176 @@
+
+* Here is an example of a rate limit implemented by the application developer that limits the total requests made
+to a specific route to safeguard health of internal application components. In this case, no specific `headers` match
+is specified, and the rate limit is applied to all traffic flows accepted by this `HTTPRoute`.
+
+```yaml
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: RateLimitFilter
+metadata:
+  name: ratelimit-all-requests
+spec:
+  type: Global
+  rules:
+  - matches:
+    - limit:
+        requests: 1000
+	unit: Second
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: example
+spec:
+  parentRefs:
+    - name: eg
+  hostnames:
+    - www.example.com
+  rules:
+    - matches:
+        - path:
+            type: PathPrefix
+            value: /foo
+      filters:
+        - type: ExtensionRef
+          extensionRef:
+            group: gateway.envoyproxy.io
+            kind: RateLimitFilter
+            name: ratelimit-all-requests
+      backendRefs:
+        - name: backend
+          port: 3000
+```
+
+### Rate limit per distinct value
+
+* Here is an example of a rate limit implemented by the application developer to limit any unique user
+by matching on a custom `x-user-id` header. Here, user A (recognised from the traffic flow using the header
+`x-user-id` and value `a`) will be rate limited at 10 requests/hour and so will user B 
+(recognised from the traffic flow using the header `x-user-id` and value `b`).
+
+```yaml
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: RateLimitFilter
+metadata:
+  name: ratelimit-per-user
+spec:
+  type: Global
+  rules:
+  - matches:
+    - Type: Distinct
+      header:
+        name: x-user-id
+      limit:
+        requests: 10
+	unit: Hour
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: example
+spec:
+  parentRefs:
+    - name: eg
+  hostnames:
+    - www.example.com
+  rules:
+    - matches:
+        - path:
+            type: PathPrefix
+            value: /foo
+      filters:
+        - type: ExtensionRef
+          extensionRef:
+            group: gateway.envoyproxy.io
+            kind: RateLimitFilter
+            name: ratelimit-per-user 
+      backendRefs:
+        - name: backend
+          port: 3000
+```
+
+## Design Decisions
+
+* The initial design uses an Extension filter to apply the Rate Limit functionality on a specific HTTPRoute.
+This was preferred over the PolicyAttachment Extension mechanism, because it is unclear whether Rate Limit
+will be required to be enforced or overridden by the platform administrator or not.
+* The Rate limits are applied across all backends within a HTTPRoute, and are not applied per backend.
+* The HTTPRoute API has a `matches` field within each `rule` to select a specific traffic flow to be routed to
+the destination backend. The RateLimitFilter API that can be attached to an HTTPRoute via an `extensionRef` filter,
```

Comment:
and the `filters` field of an HTTPRoute rule defines the "filters" that are applied to requests that match this rule. I'm trying to understand why matches of an HTTProute can't be used instead of matches for RateLimitFilter. For example:
```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: example
spec:
  ...
  rules:
    - matches:
        - path:
            type: PathPrefix
            value: /foo
      filters:
        - type: ExtensionRef
          extensionRef:
            group: gateway.envoyproxy.io
            kind: RateLimitFilter
            name: ratelimit-per-user 
      backendRefs:
        - name: backend
          port: 3000
```

Will rate limit all traffic to `/foo`. Add additional matches, e.g. headers, to be more specific on the matching criteria. Create another rule, e.g. route all other traffic after the above rule to route the rest of the traffic without rate limiting. From a UX standpoint, I like using HTTPRoute rules for match, filter, and route to backend and using the RateLimitFilter to describe the rate limiting, e.g. type local/global, descriptors, burst (local), etc.

Reply:
@danehans ptal at the Design Decision section, Ive elaborated on the need to add 2 `matches` within each API. 
Here's an example trying to make it more clear. 

Goal - app developer trying to ratelimit a specific client who can be identified using the `x-user-id` header and has the value `foo` for all their current L7 route

1) Here is the recommended way to do this

* The app dev describes a `RateLimitFilter` with the exact intent (`what` the limit is and for whom) and applies it
```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: RateLimitFilter
metadata:
  name: ratelimit-specific-user
spec:
  type: Global
  rules:
  - matches:
    - header:
        name: x-user-id
	value: foo
      limit:
        requests: 10
	unit: Hour
```	

* The app dev would have already created a route earlier to allow external traffic in, based on some match condition.
Now they would edit this route and link the rate limit filter to it.
HTTPRoute
```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: business-logic-route
spec:
  ...
  rules:
    - matches:
        - path:
            type: PathPrefix
            value: /business-logic-path
      filters:
        - type: ExtensionRef
          extensionRef:
            group: gateway.envoyproxy.io
            kind: RateLimitFilter
            name: ratelimit-specifc-user 
      backendRefs:
        - name: backend
          port: 3000
```

2) Designing above intent in a world where `matches` doesnt exist in `RateLimitFilter`

Problem 1 (sprinking domain intent in multiple places)
When the app developer writes intent for a ratelimiting case, the information will be divided in two places - the `limit` will live in the `RateLimitFilter` config but the "for whom" will live in another config
```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: RateLimitFilter
metadata:
  name: ratelimit-specific-user
spec:
  type: Global
  rules:
  - limit:
       requests: 10
       unit: Hour
```	

Problem 2 (duplicating routing intent in a single file)
Now the app developer has to duplicate the routing match section and create two similar `matches`, one for the non ratelimited route and one for the ratelimited route.
```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: business-logic-route
spec:
  ...
  rules:
    - matches:
        - path:
            type: PathPrefix
            value: /business-logic-path
    - matches:
        - path:
            type: PathPrefix
            value: /business-logic-path
    - header:
        name: x-user-id
	value: foo            
      filters:
        - type: ExtensionRef
          extensionRef:
            group: gateway.envoyproxy.io
            kind: RateLimitFilter
            name: ratelimit-specifc-user             
      backendRefs:
        - name: backend
          port: 3000
```          

* not a big issue but above presentation causes suboptimal xds translations (duplicated route matches)

Reply:
I don't see ^ as a duplication. IMO having `matches` in 1 location provides an easier-to-understand mental model. Users will be familiar with route concepts, i.e. match, filter, and forward. The above example can be reorganized and easily explained as match a path + header, rate limit the matched requests using the rules defined in `ratelimit-specifc-user`, and forward requests within the create limit to Service `backend`. Per [Gateway API docs](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteRule), all other traffic will be matched as `/` and can be routed w/o any filters to a different backend. This can be considered the "default" route and is similar in concept to network-level routing. Any additional request routing is expressed with additional match, filter, and forward (backendRefs) rules.

As stated in https://github.com/envoyproxy/gateway/pull/706#discussion_r1051165778, I'm +1 on not including "submatching" in RateLimitFilter at this time. In the future, if users provide feedback that expressing all `match` logic for rate limiting in routes is suboptimal, we can consider adding `matches` to `ratelimitfilter.spec.rules[]`.

Reply:
@danehans another reason to not reuse `matches` defined in the `HTTPRoute`, it is designed to match traffic meant for a backed / API endpoint and for such a case, higher level attributes such as `header` , `queryParam` are used.
For ratelimiting, attributes identifying clients need to be matched, they could be higher level attributes such as `header`, but also lower level attributes such as IP `subnet` . This isn't added in the initial design (to keep scope small), but will be added soon (and has already been discussed in this [PR](https://github.com/envoyproxy/gateway/pull/706#discussion_r1015806182))

Reply:
jwt claim is also a popular way to add client info for ratelimiting ([here's](https://www.docker.com/blog/checking-your-current-docker-pull-rate-limits-and-status/) an eexample)
such cases will need its own match field